### PR TITLE
Add a Container#stop method to exit early and send a finished event

### PR DESCRIPTION
**Pull request by dblandin**
This PR allows the running container to stop analysis immediately in case of an output processing error.

### CODE REVIEW

File: lib/cc/analyzer/container.rb [hunk 1]
```diff
@@ -67,6 +67,12 @@ def run(options = [])
         end
       end
 
+      def stop
+        @on_output = ->(*) { }
+        Process.kill("TERM", @pid)
```
Comment (GordonDiggs) [inline lib/cc/analyzer/container.rb]: What happens if `@pid` is nil here?

Comment (dblandin) [inline lib/cc/analyzer/container.rb]: Looks like if `@pid` were `nil` we would get a `TypeError: no implicit conversion from nil to integer` but I don't think that's possible. `Container#stop` is only called in response to output emitted from the running engine which would have been assigned a PID. If the process terminates before we send a `TERM` command, we'll get an `Errno::ESRCH` exception, which we're already handling with a no-op rescue.

File: lib/cc/analyzer/container.rb [hunk 2]
```diff
@@ -67,6 +67,13 @@ def run(options = [])
         end
       end
 
+      def stop
+        @on_output = ->(*) { }
```
Comment (pbrisbin) [inline lib/cc/analyzer/container.rb]: Why is this necessary?

File: lib/cc/analyzer/container.rb [hunk 3]
```diff
@@ -67,6 +67,13 @@ def run(options = [])
...
+        Process.kill("TERM", @pid)
+      rescue Errno::ESRCH
+        Analyzer.logger.info("Tried to send TERM to process with PID #{@pid}, but not found.")
```
Comment (pbrisbin) [inline lib/cc/analyzer/container.rb]: Probably `warn`? I'm also OK completely ignoring, maybe with a statsd increment.

File: lib/cc/analyzer/container.rb (hunk 2 above)
Comment (dblandin) [inline lib/cc/analyzer/container.rb]: It seems like the process (and output) still runs for a bit even after sending `TERM` or `KILL` so this prevents a bunch of extra errored events from reaching Kafka and triggering unprocessable state changes on the snapshot.

File: lib/cc/analyzer/container.rb (hunk 3 above)
Comment (dblandin) [inline lib/cc/analyzer/container.rb]: I was getting a Code Climate issue for leaving this empty. I think a statsd increment sounds good.

File: lib/cc/analyzer/container.rb (hunk 1 above)
Comment (GordonDiggs) [inline lib/cc/analyzer/container.rb]: I think this should be resilient to being called prematurely. Would a simple `if @pid` achieve that?

File: lib/cc/analyzer/engine.rb [hunk 4]
```diff
@@ -33,7 +33,7 @@ def run(stdout_io, container_listener)
 
         container.on_output("\0") do |output|
           unless output_filter.filter?(output)
-            stdout_io.write(output)
+            container.stop unless stdout_io.write(output)
```
Comment (GordonDiggs) [inline lib/cc/analyzer/engine.rb]: this trailing conditional seems odd to me because `sdtout_io.write` is the primary action here. What do you think of

```ruby
stdout_io.write(output) or container.stop
```

I don't usually advocate using `or` but given that it's for control flow and not logic, I think it's a good use case.

File: lib/cc/analyzer/container.rb (hunk 1 above)
Comment (dblandin) [inline lib/cc/analyzer/container.rb]: Sure, that works for me :+1:

File: lib/cc/analyzer/engine.rb (hunk 4 above)
Comment (dblandin) [inline lib/cc/analyzer/engine.rb]: Yep, I think that gives more weight to the primary action. :+1:

File: lib/cc/analyzer/container.rb [hunk 5]
```diff
@@ -67,6 +67,13 @@ def run(options = [])
         end
       end
 
+      def stop
+        @on_output = ->(*) { }
+        Process.kill("TERM", @pid)
+      rescue Errno::ESRCH
+        Analyzer.statsd.increment("container.kill_process_rescue")
```
Comment (dblandin) [inline lib/cc/analyzer/container.rb]: @pbrisbin Is this fine as a metric key or would you prefer something else?

**Comment by GordonDiggs** [comment]
LGTM

File: lib/cc/analyzer/engine.rb (hunk 4 above)
Comment (pbrisbin) [inline lib/cc/analyzer/engine.rb]: > I don't usually advocate using or but given that it's for control flow and not logic, I think it's a good use case.

FWIW, I love `and` / `or` for control flow. That's what they're there for; it's why the lower precedence.

```rb
foo_bar or raise "x"

x = get_it and go(x)
```

We seem to avoid them here, so I've been abiding but I'd be happy to have a discussion about bringing them back into our toolbox.

**Comment by pbrisbin** [comment]
LGTM too
